daemon/src/rollover/taker.rs

```diff
@@ -0,0 +1,223 @@
+use crate::rollover::protocol::DialerMessage;
+use crate::rollover::protocol::ListenerMessage;
+use crate::rollover::protocol::Propose;
+use crate::setup_contract;
+use anyhow::Context;
+use async_trait::async_trait;
+use futures::future;
+use futures::SinkExt;
+use futures::StreamExt;
+use maia_core::secp256k1_zkp::schnorrsig;
+use model::libp2p::PeerId;
+use model::OrderId;
+use model::Role;
+use model::Timestamp;
+use std::time::Duration;
+use tokio_tasks::Tasks;
+use xtra::message_channel::MessageChannel;
+use xtra::Address;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::OpenSubstream;
+use xtra_productivity::xtra_productivity;
+
+/// The duration that the taker waits until a decision (accept/reject) is expected from the maker
+///
+/// If the maker does not respond within `DECISION_TIMEOUT` seconds then the taker will fail the
+/// rollover.
+const DECISION_TIMEOUT: Duration = Duration::from_secs(30);
+
+/// One actor to rule all the rollovers
+pub struct Actor {
+    endpoint: Address<Endpoint>,
+    oracle_pk: schnorrsig::PublicKey,
+    get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+    n_payouts: usize,
+    tasks: Tasks,
+    executor: command::Executor,
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[derive(Copy, Clone)]
+pub struct ProposeRollover {
+    pub order_id: OrderId,
+    pub maker_peer_id: PeerId,
+}
+
+impl Actor {
+    pub fn new(
+        endpoint: Address<Endpoint>,
+        executor: command::Executor,
+        oracle_pk: schnorrsig::PublicKey,
+        get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+        n_payouts: usize,
+    ) -> Self {
+        Self {
+            endpoint,
+            tasks: Tasks::default(),
+            executor,
+            get_announcement,
+            oracle_pk,
+            n_payouts,
+        }
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    pub async fn handle(&mut self, msg: ProposeRollover) {
+        let ProposeRollover {
+            order_id,
+            maker_peer_id,
+        } = msg;
+
+        let substream = match self
+            .endpoint
+            .send(OpenSubstream::single_protocol(
+                maker_peer_id.inner(),
+                rollover::PROTOCOL,
+            ))
+            .await
+            .context("Endpoint is disconnected")
+            .context("Failed to open substream")
+        {
+            Ok(Ok(substream)) => substream,
+            // TODO: Same error path and add failed
+            Ok(Err(e)) => {
+                tracing::error!("Failed to open connection to maker: {e:#}");
+                return;
+            }
+            Err(e) => {
+                tracing::error!(%order_id, "Failed to open connection to maker for  rollover: {e:#}");
+
+                if let Err(e) = self
+                    .executor
+                    .execute(order_id, |cfd| Ok(cfd.fail_rollover(e)))
```

Comment:
❓ Why record a failure? Surely nothing has happened yet in the database at this stage.

Reply:
We still have the paradigm of `started` and `failed`. If we only have a `started` event then the logic in `model::Cfd` won't allow triggering a rollover, because it assumes that one is still ongoing. Only if we add `completed`, `rejected` or `failed` the rollover is "finished". 
I don't like this, but I would not change this logic as part of this work.

Reply:
recording a failure is trace of event that happened (and was significant). can be useful for analysing trends in grafana. 
and as @da-kami pointed out above, we actually currently need to do that to accurately determine ongoing rollover for a Cfd